### PR TITLE
Contributor advice

### LINKED ISSUE #752: Contributor advice proposal
I see the need to collect some advice how to build good Alpakka modules following the style that has evolved.
Please comment to get a common view of this. I'll create a PR with it when we get close to the next release.

Here is goes:

# Some advice for Alpakka contributors

## API design

Depending on the technology you integrate with Akka Streams and Alpakka you'll create Sources, Flows and Sinks.

When designing Sinks, please check if a Flow might be a better fit. Most other systems will give some kind of response which
could continue in a Flow. Sinks can easily created be attaching a Sink.ignore to the Flow. In case you want to offer 
Sinks based on Flows, focus on the most important cases and leave the more intricate variants just in the Flow API.

When designing Flows, consider adding an extra field to the in- and out-messages which is passed through. A common
use case we see, is committing a Kafka offset after passing data to another system.

### Settings

Most technologies will have a couple of configuration settings that will be needed for all Flows. Create case classes
collecting these settings instead of passing them in every method.

Create a default instance in the companion object with good defaults which can be updated via `copy` or `withXxxx` methods.

As Java can't use the `copy` methods, add `withXxxx` methods to specify certain fields in the settings instance. 

In case you see the need to support reading the settings from `Config`, offer a method taking the `Config` instance so
that the user can apply a proper namespace.
Refrain from using `akka.stream` as prefix, prefer `alpakka` as root namespace.

### Packages & Scoping

Use `private`, `private[connector]` and `final` extensively to limit the API surface. 

| Package                                  | Purpose
| -----------------------------------------|------------------------
| `akka.stream.alpakka.connector.javadsl`  | Java-only part of the API, normally factories for Sources, Flows and Sinks
| `akka.stream.alpakka.connector.scaladsl` | Scala-only part of the API, normally factories for Sources, Flows and Sinks
| `akka.stream.alpakka.connector`          | Shared API, eg. settings classes, implementation if thorowly made private
| `akka.stream.alpakka.connector.impl`     | Optional, implementation in separate package

### Regarding Java APIs

* Provide factory methods for Sources, Flows and Sinks in the `javadsl` package
* The Akka Stream Scala instances have a `.asJava` method to convert to the `akka.stream.javadsl` counterparts
* Let the Java API use Java standard types (`java.util.Optional`, `java.util.List`, ...) and convert using 
`JavaConverters` where needed
* Use the `akka.japi.Pair` class to return tuples
* Use `scala.compat.java8.FutureConverters` to translate Futures to `CompletionStage`s

## Graph stage checklist

* Keep mutable state within the `GraphStageLogic` only
* Open/close connections in `preStart` 
* Release resources in `postStop`
* Fail early on configuration errors

## Documentation

Create or complement the documentation in the `docs` module using [Paradox](https://github.com/lightbend/paradox) syntax.
Prepare code snippets to be integrated by Paradox in the tests.

Use ScalaDoc if you see the need to describe the API usage better than the naming does.

### PR DESCRIPTION
**Pull request by ennru**
Fixes #752

### CODE REVIEW

File: contributor-advice.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+Write tests for all public methods and possible settings.
+
+Use Docker containers for any testing needed by your application. The container's configuration can be added to the
+`docker-compose.yml` file.
+Please ensure that you limit the amount of resources used by the containers.
+
+## Documentation
+
+Using [Paradox](https://github.com/lightbend/paradox) syntax, create or complement the documentation in the `docs` module.
```
Comment (ktoso) [inline contributor-advice.md]: syntax (which is very close to markdown)

File: contributor-advice.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+Prepare code snippets to be integrated by Paradox in the tests. Such example should be part of real tests and no in
```
Comment (ktoso) [inline contributor-advice.md]: no => not

File: contributor-advice.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+* Make sure the code is thread-safe
+
+### Keep the code DRY
+
+Avoid duplication of code between different Sources, Sinks and Flows. Extract the common logic to a common abstract
+base `GraphStageLogic` that is inherited by the `GraphStage`s.
+
+In some cases it may be useful to implement the Sources or the Sinks by means of Flows:
+* Source: `Source.maybe.viaMat(MyFlow)(Keep.right)`
+* Sink: `MyFlow.toMat(Sink.ignore)(Keep.right)`
```
Comment (ktoso) [inline contributor-advice.md]: I'd reword a bit: Sometimes it may be useful for users to provide them with Sink or Source versions of an operation, even if the main concept is implemented (and most useful) as a Flow. This can be easily implemented as (reusing the core Flow implementation):

File: contributor-advice.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+| `akka.stream.alpakka.connector.scaladsl` | Scala-only part of the API, normally factories for Sources, Flows and Sinks
+| `akka.stream.alpakka.connector`          | Shared API, eg. settings classes, implementation if thorowly made private
+| `akka.stream.alpakka.connector.impl`     | Optional, implementation in separate package
+
+### Graph stage checklist
+
+* Keep mutable state within the `GraphStageLogic` only 
+* Open connections in `preStart`
+* Release resources in `postStop`
+* Fail early on configuration errors
```
Comment (ktoso) [inline contributor-advice.md]: fail using `failStage` instead of throwing exceptions?

cc @patriknw

File: contributor-advice.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+* No Blocking At Any Time
```
Comment (ktoso) [inline contributor-advice.md]: yeap, classic motto, but let's explain it too, for new people:

"No Blocking At Any Time" -- in other words, avoid blocking whenever possible and replace it with asynchronous programming (async callbacks, stage actors)

File: contributor-advice.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+| `akka.stream.alpakka.connector.impl`     | Optional, implementation in separate package
+
+### Graph stage checklist
+
+* Keep mutable state within the `GraphStageLogic` only 
+* Open connections in `preStart`
+* Release resources in `postStop`
+* Fail early on configuration errors
+* No Blocking At Any Time 
+* Make sure the code is thread-safe
```
Comment (ktoso) [inline contributor-advice.md]: add: "; if in doubt, please ask!"

File: contributor-advice.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+Refrain from using `akka.stream` as prefix, prefer `alpakka` as root namespace.
+
+### Public factory methods
+
+Depending on the technology you integrate with Akka Streams and Alpakka you'll create Sources, Flows and Sinks.
+Regardless on how they are implemented make sure that you create the relevant Sources, Sinks and Flows APIs so they are
+simple and easy to use.
+
+### Java APIs
+
```
Comment (ktoso) [inline contributor-advice.md]: Alpakka, same as Akka, aims to keep 100% feature parity between the various language DSLs.

File: contributor-advice.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+* Provide factory methods for Sources, Flows and Sinks in the `javadsl` package wrapping all the methods in the Scala API
+* The Akka Stream Scala instances have a `.asJava` method to convert to the `akka.stream.javadsl` counterparts
+* Let the Java API use Java standard types (`java.util.Optional`, `java.util.List`, ...) and convert using
+`JavaConverters` where needed
+* Use the `akka.japi.Pair` class to return tuples
+* Use `scala.compat.java8.FutureConverters` to translate Futures to `CompletionStage`s
```
Comment (ktoso) [inline contributor-advice.md]: also mention collection converters? Suggest using List in Java

File: contributor-advice.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
+# Some advice for Alpakka contributors
+
+## API design
+
+### Flows
+
+When designing Flows, consider adding an extra field to the in- and out-messages which is passed through. A common
+use case we see, is committing a Kafka offset after passing data to another system.
+
+### Settings
```
Comment (ktoso) [inline contributor-advice.md]: move settings under javadsl since we talk about it here

File: contributor-advice.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+In case you see the need to support reading the settings from `Config`, offer a method taking the `Config` instance so
+that the user can apply a proper namespace.
+Refrain from using `akka.stream` as prefix, prefer `alpakka` as root namespace.
+
+### Public factory methods
+
+Depending on the technology you integrate with Akka Streams and Alpakka you'll create Sources, Flows and Sinks.
+Regardless on how they are implemented make sure that you create the relevant Sources, Sinks and Flows APIs so they are
+simple and easy to use.
```
Comment (ktoso) [inline contributor-advice.md]: good

File: contributor-advice.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+### Packages & Scoping
+
+Use `private`, `private[connector]` and `final` extensively to limit the API surface.
+
+| Package                                  | Purpose
+| -----------------------------------------|------------------------
+| `akka.stream.alpakka.connector.javadsl`  | Java-only part of the API, normally factories for Sources, Flows and Sinks
+| `akka.stream.alpakka.connector.scaladsl` | Scala-only part of the API, normally factories for Sources, Flows and Sinks
+| `akka.stream.alpakka.connector`          | Shared API, eg. settings classes, implementation if thorowly made private
+| `akka.stream.alpakka.connector.impl`     | Optional, implementation in separate package
```
Comment (ktoso) [inline contributor-advice.md]: great table

File: contributor-advice.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
+# Some advice for Alpakka contributors
+
+## API design
+
+### Flows
```
Comment (juanjoDiaz) [inline contributor-advice.md]: I'd move this whole `Flows` section below the `Public factory methods` where we mention the Sources, Sinks and Flows.

File: contributor-advice.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+When designing Flows, consider adding an extra field to the in- and out-messages which is passed through. A common
+use case we see, is committing a Kafka offset after passing data to another system.
+
+### Settings
+
+Most technologies will have a couple of configuration settings that will be needed for all Flows. Create case classes
```
Comment (juanjoDiaz) [inline contributor-advice.md]: Most technologies will have a couple of configuration settings that will be needed for all *connector*

We do more than flows here :)

File: contributor-advice.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+Most technologies will have a couple of configuration settings that will be needed for all Flows. Create case classes 
+collecting these settings instead of passing them in every method.
+
+Create a default instance in the companion object with good defaults which can be updated via `copy` or `withXxxx` methods.
+
+As Java can't use the `copy` methods, add `withXxxx` methods to specify certain fields in the settings instance.
+
+In case you see the need to support reading the settings from `Config`, offer a method taking the `Config` instance so
+that the user can apply a proper namespace.
+Refrain from using `akka.stream` as prefix, prefer `alpakka` as root namespace.
```
Comment (juanjoDiaz) [inline contributor-advice.md]: This sentence is very true but has little to do with Settings. I'd move it to a separate section for this kind of general stuff

File: contributor-advice.md (hunk 6 above)
Comment (juanjoDiaz) [inline contributor-advice.md]: I'd move the thread-safe bullet point to be before the no-blocking one. Since blocking tends to be the first approach that people think to tackle thread safety.

File: contributor-advice.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+* Sink: `MyFlow.toMat(Sink.ignore)(Keep.right)`
+
+You do not need to expose every configuration a Flow offers this way -- Focus on the most important ones.
+
+
+## Test
+
+Write tests for all public methods and possible settings.
+
+Use Docker containers for any testing needed by your application. The container's configuration can be added to the
```
Comment (burakkose) [inline contributor-advice.md]: In my view, before suggesting to use docker, we should encourage to use internal testing API like MiniSolrCloudCluster for Solr, ElasticsearchClusterRunner for Elasticsearch, if it is possible.

Comment (juanjoDiaz) [reply contributor-advice.md]: Using Docker make the testing environment more isolated than if we configure stuff just for the test. It allows us to limit, memory, disk or whatever we want to limit. And it's trivial to update regardless of the connector version or dependencies.

Also, I'd rather pulling a container for testing than adding an extra dependency.

File: contributor-advice.md (hunk 8 above)
Comment (ennru) [reply contributor-advice.md]: Anything more than it says in "Let the Java API use Java standard types (`java.util.Optional`, `java.util.List`, ...) and convert using `JavaConverters` where needed" ?

File: contributor-advice.md (hunk 14 above)
Comment (ennru) [reply contributor-advice.md]: Ah, it talks about the Config namespace. No the package names. I make it more specific.

File: contributor-advice.md (hunk 8 above)
Comment (ktoso) [reply contributor-advice.md]: Sounds good

Comment (ktoso) [reply contributor-advice.md]: Good as is then

**Comment by juanjoDiaz** [comment]
Does that last commit about Slick belong to this PR or was it accidental?

**Comment by ennru** [comment]
It does not belong to it, but was intentional. The builds failed with too long log output, so I need to cut it down.
